File: simtix/src/cores/pipelined/lsu/stack_remap_table.cc
/*
 * SPDX-FileCopyrightText: 2026 CASLab, National Cheng Kung University
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include "cores/pipelined/lsu/stack_remap_table.h"

#include <liblv/binding.h>
#include <liblv/output.h>

#include <cstring>

namespace simtix::pipelined {

namespace {

bool IsPowerOfTwo(uint64_t value) {
  return value != 0 && (value & (value - 1)) == 0;
}

}  // namespace

StackRemapTable::StackRemapTable(const sc_module_name &name, const Param &param)
    : sc_module(name),
      region_mask_(~(param.region_size - 1)),
      descriptors_(param.entries, 0),
      sink_("mmio_port") {
  if (param.entries == 0) {
    LV_FATAL("StackRemapTable must contain at least one entry");
  }
  if (!IsPowerOfTwo(param.region_size)) {
    LV_FATAL("StackRemapTable region_size must be a power of two, got {}",
             param.region_size);
  }

  SC_THREAD(HandleRequests);
}

bool StackRemapTable::IsDescriptorValid(uint64_t descriptor) const {
  return (descriptor & kValidBitMask) != 0;
}

uint64_t StackRemapTable::NormalizeDescriptor(uint64_t descriptor) const {
  return (descriptor & kAddressMask & region_mask_) |
         (descriptor & kValidBitMask);
}

bool StackRemapTable::Matches(uint64_t addr) const {
  const uint64_t region_base = addr & region_mask_;
  for (uint64_t descriptor : descriptors_) {
    if (IsDescriptorValid(descriptor) &&
        (descriptor & region_mask_) == region_base) {
      return true;
    }
  }
  return false;
}

void StackRemapTable::HandleRequests() {
  for (;;) {
    auto *trans = sink_.req_port->read();
    const uint64_t addr = trans->get_address();
    const unsigned int size = trans->get_data_length();

    if (size != sizeof(uint64_t) || addr % sizeof(uint64_t) != 0 ||
        addr + size > descriptors_.size() * sizeof(uint64_t)) {
      trans->set_response_status(tlm::TLM_ADDRESS_ERROR_RESPONSE);
      sink_.resp_port->write(trans);
      continue;
    }

    uint64_t &descriptor = descriptors_[addr / sizeof(uint64_t)];
    if (trans->is_write()) {
      uint64_t value = 0;
      std::memcpy(&value, trans->get_data_ptr(), sizeof(value));
      descriptor = NormalizeDescriptor(value);
    } else if (trans->is_read()) {
      std::memcpy(trans->get_data_ptr(), &descriptor, sizeof(descriptor));
    } else {
      trans->set_response_status(tlm::TLM_COMMAND_ERROR_RESPONSE);
      sink_.resp_port->write(trans);
      continue;
    }

    trans->set_response_status(tlm::TLM_OK_RESPONSE);
    sink_.resp_port->write(trans);
  }
}

LV_BINDING(simtix, StackRemapTable)
    .constructor(
        [](const char *name, const StackRemapTable::Param &param) {
          return std::make_shared<StackRemapTable>(name, param);
        },
        lv::params("name", "param"),
        lv::doc("Create a firmware-programmable stack remapping table"))
    .property("mmio_port", &StackRemapTable::port,
              lv::doc("Stack remapping table MMIO target"))
    .method("matches", &StackRemapTable::Matches, lv::params("address"),
            lv::doc("Return whether an address matches a valid descriptor"));

}  // namespace simtix::pipelined

File: simtix/tests/pipelined_core/stack_remap_table.lua
-- SPDX-FileCopyrightText: 2026 CASLab, National Cheng Kung University
--
-- SPDX-License-Identifier: Apache-2.0

local period = sc.time(1, sc.time_unit.NS)
local clock = sc.clock("clock", period)
local initiator = simple.Initiator("initiator")
local table = simtix.StackRemapTable("stack_remap_table", {
  entries = 4,
  region_size = 0x10000,
})

initiator.clock = clock
initiator.target = table.mmio_port

local function u64_le(value)
  local bytes = {}
  for i = 1, 8 do
    bytes[i] = value % 256
    value = math.floor(value / 256)
  end
  return bytes
end

local function wait_for_completion(target)
  for _ = 1, 32 do
    if initiator:completed_count() >= target then return end
    sc.start(period)
  end
  error("Timed out waiting for stack remap MMIO transaction")
end

local stack_base = 0x81000000
assert(not table:matches(stack_base))

local target = initiator:completed_count() + 1
initiator:add_payload({
  addr = 0,
  data = u64_le(stack_base + 1),
})
wait_for_completion(target)

assert(table:matches(stack_base))
assert(table:matches(stack_base + 0xFFFF))
assert(not table:matches(stack_base + 0x10000))
assert(not table:matches(0x82000000))

target = initiator:completed_count() + 1
initiator:add_payload({
  addr = 0,
  data = u64_le(0),
})
wait_for_completion(target)
assert(not table:matches(stack_base))
